### PR TITLE
virtualenv links against wrong python #978

### PR DESCRIPTION
**Pull request by themperek**
Addressing #978. It seems cleaner.

PS. In case this gets accepted please squash.

### CODE REVIEW

**Comment by imphil** [comment]
Thanks for this improvement @themperek, looks cleaner indeed. 

However, I'm a bit worried about the performance impact of calling Python multiple times. Can we get away with calling PYTHON_BIN once, and then letting make split the result into the individual variables?

**Comment by themperek** [comment]
>Can we get away with calling PYTHON_BIN once, and then letting make split the result into the individual variables?

We can but then we need to parse this with some Unix tools and this will be ugly (in my opinion).

**Comment by eric-wieser** [comment]
What if we invoked python once, and had it output makefile syntax? `$(eval $(shell cocotb-config get-makefile-vars))`

**Comment by themperek** [comment]
Seems like we still (how long @imphil ?) want to be able to use cocotb without cocotb-config?

File: cocotb/share/makefiles/Makefile.pylib.Darwin [hunk 1]
```diff
@@ -54,20 +49,7 @@ ifneq ($(NATIVE_ARCH),$(ARCH))
...
-else ifneq (, $(shell which $(PYTHON_BIN)$(PYTHON_MAJOR_VERSION)-config 2>/dev/null))
-	PY_CFG_EXE=$(PYTHON_BIN)$(PYTHON_MAJOR_VERSION)-config
-else ifneq (, $(shell which $(PYTHON_BIN)$(PYTHON_VERSION)-config 2>/dev/null))
-	PY_CFG_EXE=$(PYTHON_BIN)$(PYTHON_VERSION)-config
-else
-	$(error ERROR: Unable to find a valid python-config.)
-endif
-
-PYLIBS = $(shell $(PY_CFG_EXE) --libs)
+PYLIBS:=$(shell $(PYTHON_BIN) -c 'from __future__ import print_function; from distutils import sysconfig; import os; print("-l"+os.path.splitext(sysconfig.get_config_var("LIBRARY"))[0][3:])')
```
Comment (eric-wieser) [inline cocotb/share/makefiles/Makefile.pylib.Darwin]: According to the implementation of `python-config.py`, the implementation should be something like:
```
    elif opt in ('--libs', '--ldflags'):
        libs = []
        if '--embed' in opt_flags:
            libs.append('-lpython' + pyver + sys.abiflags)
        else:
            libpython = getvar('LIBPYTHON')
            if libpython:
                libs.append(libpython)
        libs.extend(getvar('LIBS').split() + getvar('SYSLIBS').split())

        # add the prefix/lib/pythonX.Y/config dir, but only if there is no
        # shared library in prefix/lib/.
        if opt == '--ldflags':
            if not getvar('Py_ENABLE_SHARED'):
                libs.insert(0, '-L' + getvar('LIBPL'))
        print(' '.join(libs))
```

Comment (themperek) [reply cocotb/share/makefiles/Makefile.pylib.Darwin]: Maybe we should wait with all those changes until we can use `cocotb-config`.

**Comment by imphil** [comment]
Make can do word-splitting natively with `$(word n,text)` (https://www.gnu.org/software/make/manual/html_node/Text-Functions.html), with one important limitation: spaces spaces in paths are not supported. This is a general make limitation, not only for this function.

However, it feels to me that the "no spaces in paths" rule wasn't as important until now as most paths to source files are relative. With absolute paths and Windows added to the mix it's a whole different story.

So:
- I'd take this patch as-is, as it strictly improves things.
- Open a follow-up trying to find a sane way to avoid these multiple calls to an external tool, be it PYTHON_BIN or cocotb-config. (It's the same overhead.)

**Comment by imphil** [comment]
> Seems like we still (how long @imphil ?) want to be able to use cocotb without cocotb-config?

We can require cocotb-config. We just cannot require it being installed into PATH. Instead, we need to look it up from the old COCOTB environment variable. (I will do that in #990 when I find time for it, hopefully in the next couple of days.)

**Comment by themperek** [comment]
Alternatively, we go into the direction to generate some makefile with cocotb-config with all this configuration handled in python?

**Comment by themperek** [comment]
It looks to me it is OK. And seems to work across diffrent build environments. 
The objection I would have is that "LDLIBRARY"/"LIBRARY" variable is not really documented although a GitHub search suggests this is widely used.

**Pull request merged**